src/libsync/propagatedownloadzsync.cpp

```diff
@@ -0,0 +1,320 @@
+#include "filesystem.h"
+#include "propagatorjobs.h"
+#include "common/checksums.h"
+#include "common/asserts.h"
+
+#include <QLoggingCategory>
+#include <QNetworkAccessManager>
+#include <QFileInfo>
+#include <QDir>
+#include <QTemporaryFile>
+#include <cmath>
+
+#ifdef Q_OS_UNIX
+#include <unistd.h>
+#endif
+
+namespace OCC {
+
+// DOES NOT take ownership of the device.
+GETFileZsyncJob::GETFileZsyncJob(AccountPtr account, SyncFileItemPtr &item, OwncloudPropagator *propagator,
+    const QString &remotepath, const QString &localpath, const QString &localtmppath, QFile *device,
+    const QMap<QByteArray, QByteArray> &headers, const QByteArray &expectedEtagForResume,
+    const QByteArray &zsyncData, QObject *parent)
+    : GETJob(account, remotepath, parent)
+    , _device(device)
+    , _item(item)
+    , _propagator(propagator)
+    , _headers(headers)
+    , _expectedEtagForResume(expectedEtagForResume)
+    , _hasEmittedFinishedSignal(false)
+    , _zsyncData(zsyncData)
+    , _zs(NULL)
+    , _zr(NULL)
+    , _localpath(localpath)
+    , _localtmppath(localtmppath)
+    , _nrange(0)
+    , _zbyterange(NULL)
+    , _current(0)
+    , _pos(0)
+    , _received(0)
+{
+    _thread.setObjectName("GETFIleZsync thread");
+}
+
+void GETFileZsyncJob::startCurrentRange()
+{
+    qlonglong start = _zbyterange[_current*2];
+    qlonglong end = _zbyterange[(_current*2)+1];
+    _headers["Range"] = "bytes=" + QByteArray::number(start) + '-' + QByteArray::number(end);
+    qCDebug(lcZsyncGet) << path() << "HTTP GET with range" << _headers["Range"];
+
+    QNetworkRequest req;
+    for (QMap<QByteArray, QByteArray>::const_iterator it = _headers.begin(); it != _headers.end(); ++it) {
+        req.setRawHeader(it.key(), it.value());
+    }
+
+    req.setPriority(QNetworkRequest::LowPriority); // Long downloads must not block non-propagation jobs.
+
+    sendRequest("GET", makeDavUrl(path()), req);
+
+    reply()->setReadBufferSize(16 * 1024); // keep low so we can easier limit the bandwidth
+    qCDebug(lcZsyncGet) << _bandwidthManager << _bandwidthChoked << _bandwidthLimited;
+    if (_bandwidthManager) {
+        _bandwidthManager->registerDownloadJob(this);
+    }
+
+    if (reply()->error() != QNetworkReply::NoError) {
+        qCWarning(lcZsyncGet) << " Network error: " << errorString();
+    }
+
+    _pos = 0;
+
+    connect(reply(), &QIODevice::readyRead, this, &GETFileZsyncJob::slotReadyRead);
+    connect(reply(), &QNetworkReply::downloadProgress, this, &GETFileZsyncJob::slotDownloadProgress);
+    connect(reply(), &QNetworkReply::metaDataChanged, this, &GETFileZsyncJob::slotMetaDataChanged);
+    connect(this, &AbstractNetworkJob::networkActivity, account().data(), &Account::propagatorNetworkActivity);
+}
+
+bool GETFileZsyncJob::finished()
+{
+    if (reply()->bytesAvailable()) {
+        return false;
+    }
+
+    if (_current < _nrange - 1) {
+        _current++;
+        startCurrentRange();
+        return false;
+    }
```

Comment:
So, this chains a bunch or byte-range-GETs to the server and passes the data to zsync. What I'm still wondering is: Does it do anything to the file that was fed in the 'seed' stage? It's very important that that file is preserved as is (in case of conflicts).

Reply:
Yeah I had to resort to changing because I realized that Sabre doesn't support multi-part ranges. Pretty sure zsync is writing into a temporary file at this stage via `zsync_receive_data`, will double check.

Reply:
Confirmed, no touching of original file until done.